File: Medium/279. Perfect Squares.cpp
class Solution {
public:
	unordered_map<int,int> dict;

    int numSquares(int n) {
		if(n <= 3) return n;

        int square = int(sqrt(n));
		int minNum = n;
		for(int i=square; i>1; i--){
			int num = n - i*i;

			int value;
			if(dict.count(num) > 0){
				value = dict[num];
			}else{
				value = numSquares(n - i*i);
				dict[num] = value;
			}
			minNum = min(minNum, value);
		}

		return minNum + 1;
    }
};

File: Medium/279. Perfect Squares.py
class Solution:
    def __init__(self):
        self.dict = {}
        
    def numSquares(self, n: int) -> int:
        if n <= 3: return n
        
        square = int(n ** 0.5)
        minNum = n
        for i in range(square,1,-1):
            num = n - i**2
            if num in self.dict:
                value = self.dict[num]
            else:
                value = self.numSquares(n - i**2)
                self.dict[num] = value
            minNum = min(minNum, value)
            
        return minNum + 1